### PR TITLE
Support QR hardware

### PR DESCRIPTION
**Pull request by soralit**
**Add QR hardware support**

**Description**
- CHANGED:
  - Change `signTypedMessage` method to support QR hardware. 

- ADDED:
  - Add QR hardware Keyring 
  - Add QR hardware's interactive methods, including read keyring, submit signature and related methods.

**Checklist**

- [ ] Tests are included if applicable
- [ ] Any added code is fully documented

### CODE REVIEW

**Comment by soralit** [comment]
Hi @gantunesr , I will be on holidays next week, so I just raised this PR (which is not fully tested) to talk about the API definition and code style. Please have a look when you get time. 

Besides, I have a question about `signTypedMessage`: Why not just use the `signTypedMessage` method in `ETHKeyringController` but exporting the account and sign messages in a raw way? I think all the keyring types have implemented `signTypedData` method.

**Comment by gantunesr** [comment]
Hey @soralit, about your question regarding `signTypedMessage` and the other devs in the team can correct me if needed. Looks like the reason was mainly about the standardization around the method and the controllers. I wasn't here at that moment, so I'm not really sure about the details.

File: src/keyring/KeyringController.ts [hunk 1]
```diff
@@ -577,6 +596,94 @@ export class KeyringController extends BaseController<
...
+  connectQRHardware = async (page: number) => {
+    const keyring = await this.getQRKeyring();
+    let accounts = [];
+    switch (page) {
+      case -1:
+        accounts = await keyring.getPreviousPage();
+        break;
+      case 1:
+        accounts = await keyring.getNextPage();
+        break;
```
[the remark this answers is not in the document]
Comment (soralit) [reply src/keyring/KeyringController.ts:655]: Good catch, I will add a check here.

File: src/keyring/KeyringController.ts [hunk 2]
```diff
@@ -577,6 +596,94 @@ export class KeyringController extends BaseController<
...
+
+  /**
+   * Get qr hardware keyring.
+   *
+   * @returns The added keyring
+   */
+  private async getQRKeyring() {
+    const keyring = privates
+      .get(this)
+      .keyring.getKeyringsByType(KeyringTypes.qr)[0];
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply src/keyring/KeyringController.ts:619]: But is it possible to have more instances?

[the remark this answers is not in the document]
Comment (soralit) [reply src/keyring/KeyringController.ts:619]: Hi @gantunesr , sorry for miss this comment. The answer is no, because the API definition limit the usage of `addQRKeyring`.

File: src/keyring/KeyringController.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (soralit) [reply src/keyring/KeyringController.ts]: Hi @gantunesr , I checked the code again and if there has no previous page, it will return the first page of accounts. So I add an try/catch to catch the error about exceeding the boundary for `next page`

**Review by mcmire — COMMENTED** [review]
I took a look at this PR. I am a bit wary of these changes because it feels like it expands the responsibilities of KeyringController in a way that doesn't feel quite right. After this change, there will be some methods that are designed to work with our Keyring objects, and there will be some methods that are designed to work with yours.

That said, we know that our existing keyring code could use work. In this case, it doesn't help that our keyring modules are untyped. We plan on addressing these and other concerns in future updates, so I'm okay with this now with the knowledge that these changes are fairly self-contained and that we have plans for improvement later.

With that in mind I made some suggestions for how we could improve this code.

**Comment by soralit** [comment]
Hi @mcmire , thanks for your review and comments. I will look into your suggestions and update the code ASAP.

File: src/assets/AccountTrackerController.ts [hunk 3]
```diff
@@ -155,6 +155,29 @@ export class AccountTrackerController extends BaseController<
...
+   * Sync accounts balances with some additional addresses.
+   *
+   * @param addresses - the additional addresses, may be hardware wallet addresses.
+   * @returns accounts - current state accounts
+   */
+  syncWithAddresses = async (addresses: string[]) => {
+    this.syncAccounts();
+    const { accounts } = this.state;
+    addresses.forEach((address) => {
+      accounts[address] = { balance: '0x0' };
```
[the remark this answers is not in the document]
Comment (soralit) [reply src/assets/AccountTrackerController.ts:169]: Cool, the method `refresh` above has the same issue, I will update it also.

File: src/assets/AccountTrackerController.ts [hunk 4]
```diff
@@ -146,14 +146,37 @@ export class AccountTrackerController extends BaseController<
...
     this.syncAccounts();
-    const { accounts } = this.state;
+    const accounts = Object.assign({}, this.state.accounts);
+    for (const address in accounts) {
+      await safelyExecuteWithTimeout(async () => {
+        const balance = await query(this.ethQuery, 'getBalance', [address]);
+        accounts[address] = { balance: BNToHex(balance) };
+      });
+    }
+    this.update({ accounts: { ...accounts } });
```
Comment (mcmire) [inline src/assets/AccountTrackerController.ts:156]: Because you're cloning `this.state.accounts` above you can shorten this to:

``` suggestion
this.update({ accounts });
```

File: src/assets/AccountTrackerController.ts [hunk 5]
```diff
@@ -146,14 +146,37 @@ export class AccountTrackerController extends BaseController<
...
+      accounts[address] = { balance: '0x0' };
+    });
+
     for (const address in accounts) {
       await safelyExecuteWithTimeout(async () => {
         const balance = await query(this.ethQuery, 'getBalance', [address]);
         accounts[address] = { balance: BNToHex(balance) };
-        this.update({ accounts: { ...accounts } });
       });
     }
```
Comment (mcmire) [inline src/assets/AccountTrackerController.ts:177]: Is there a reason why you're setting all account balances to 0 and then requesting the balance of each account? Additionally you should be able to shorten this without needing to clone anything:

``` suggestion
    const accounts = await Promise.all(
      accounts.map((address) => {
        return safelyExecuteWithTimeout(async () => {
          const balance = await query(this.ethQuery, 'getBalance', [address]);
          return [address, balance];
        })
      })
    ).reduce((obj, [address, balance]) => {
      return { ...obj: [address]: balance }
    }, {});
```

File: src/keyring/KeyringController.ts [hunk 6]
```diff
@@ -421,6 +428,25 @@ export class KeyringController extends BaseController<
   ) {
     try {
       const address = normalizeAddress(messageParams.from);
+      const QRKeyring = await this.getOrAddQRKeyring();
```
Comment (mcmire) [inline src/keyring/KeyringController.ts:431]: Should this be `qrKeyring` instead of `QRKeyring`? Otherwise we may confuse this for a class of some sort.

File: src/keyring/KeyringController.ts [hunk 7]
```diff
@@ -577,6 +603,129 @@ export class KeyringController extends BaseController<
...
+   * @returns The added keyring
+   */
+  private async getOrAddQRKeyring() {
+    const keyring = privates
+      .get(this)
+      .keyring.getKeyringsByType(KeyringTypes.qr)[0];
+    if (keyring) {
+      return keyring;
+    }
+    return await this.addQRKeyring();
```
Comment (mcmire) [inline src/keyring/KeyringController.ts:630]: You can shorten this to:

``` suggestion
    return keyring ?? await this.addQRKeyring();
```

File: src/assets/AccountTrackerController.ts (hunk 5 above)
Comment (soralit) [reply src/assets/AccountTrackerController.ts:177]: Hi @mcmire , this method is for syncing accounts balance with some additional addresses. So I update the cloned accounts with the additional addresses firstly. 
But, actually the state update is not necessary here I think, I will modify this method to a pure function just return the additional addresses with synced balance.

**Comment by soralit** [comment]
Hi @mcmire , thanks for your suggestions. I will append a commit to add unit tests later.

File: src/keyring/KeyringController.ts [hunk 8]
```diff
@@ -421,6 +428,25 @@ export class KeyringController extends BaseController<
...
     try {
       const address = normalizeAddress(messageParams.from);
+      const qrKeyring = await this.getOrAddQRKeyring();
+      const qrAccounts = await qrKeyring.getAccounts();
+      if (
+        qrAccounts.findIndex(
+          (qrAddress: string) =>
+            qrAddress.toLowerCase() === address.toLowerCase(),
+        )
+      ) {
```
Comment (mcmire) [inline src/keyring/KeyringController.ts:438]: Oops, I wasn't clear earlier — this should probably be

``` suggestion
      if (
        qrAccounts.findIndex(
          (qrAddress: string) =>
            qrAddress.toLowerCase() === address.toLowerCase(),
        ) !== -1
      ) {
```

File: src/assets/AccountTrackerController.ts [hunk 9]
```diff
@@ -146,14 +146,40 @@ export class AccountTrackerController extends BaseController<
...
+    this.update({ accounts });
+  };
+
+  /**
+   * Sync accounts balances with some additional addresses.
+   *
+   * @param addresses - the additional addresses, may be hardware wallet addresses.
+   * @returns accounts - addresses with synced balance
+   */
+  syncBalanceWithAddresses = async (addresses: string[]) => {
```
Comment (mcmire) [inline src/assets/AccountTrackerController.ts:165]: Unless there is a reason, do you mind changing this to:

``` suggestion
  async syncBalanceWithAddresses(addresses: string[]) {
``

**Review by mcmire — COMMENTED** [review]
A few more things. Also — would you mind adding return types to each method? I know TypeScript can fill these in automatically but I think it's better if we are explicit as 1) it forces the author to think about what the return type _should_ be and 2) it removes one step for the reader to understand what the expected return type of the methods are.

**Comment by soralit** [comment]
Hi @mcmire , sorry for the late reply. I'm working on the bug fixing on mobile side these days. I just resolved the code issues and add types for new methods. I will append the unit tests tomorrow.

**Comment by soralit** [comment]
Hi @mcmire , just update the unit tests and rebased the code, please have a review when you have time.

File: src/keyring/KeyringController.ts [hunk 10]
```diff
@@ -157,11 +166,13 @@ export class KeyringController extends BaseController<
       syncIdentities,
       updateIdentities,
       setSelectedAddress,
+      setAccountLabel,
     }: {
       removeIdentity: PreferencesController['removeIdentity'];
       syncIdentities: PreferencesController['syncIdentities'];
       updateIdentities: PreferencesController['updateIdentities'];
       setSelectedAddress: PreferencesController['setSelectedAddress'];
+      setAccountLabel: PreferencesController['setAccountLabel'];
```
Comment (mcmire) [inline src/keyring/KeyringController.ts:175]: This will be a backward-incompatible change. Maybe this should be optional after all? Or does this have to be provided?

File: src/keyring/KeyringController.test.ts [hunk 11]
```diff
@@ -40,11 +46,16 @@ describe('KeyringController', () => {
     keyringTypes: string[];
     keyrings: Keyring[];
   };
-  const baseConfig: Partial<KeyringConfig> = { encryptor: new MockEncryptor() };
+  const additionalKeyrings = [QRKeyring];
+  const baseConfig: Partial<KeyringConfig> = {
+    encryptor: new MockEncryptor(),
+    keyringTypes: additionalKeyrings,
+  };
```
Comment (mcmire) [inline src/keyring/KeyringController.test.ts:53]: Mind adding an extra blank line after this line? I think it'll make the code look less smushed.

File: src/keyring/KeyringController.ts [hunk 12]
```diff
@@ -421,6 +433,25 @@ export class KeyringController extends BaseController<
...
+        if (
+          version !== SignTypedDataVersion.V1 &&
+          typeof messageParamsClone.data === 'string'
+        ) {
+          messageParamsClone.data = JSON.parse(messageParamsClone.data);
+        }
+        return privates
+          .get(this)
+          .keyring.signTypedMessage(messageParamsClone, { version });
+      }
```
Comment (mcmire) [inline src/keyring/KeyringController.ts:454]: Mind adding an extra blank line after this line? I think it'll make the code look less smushed.

**Review by mcmire — COMMENTED** [review]
Just a few more minor things. Everything generally looks good to me otherwise (although I'll ping some other people to get some second opinions).

File: src/keyring/KeyringController.ts [hunk 13]
```diff
@@ -577,6 +608,129 @@ export class KeyringController extends BaseController<
...
+  /**
+   * Get qr hardware keyring.
+   *
+   * @returns The added keyring
+   */
+  async getOrAddQRKeyring(): Promise<QRKeyring> {
+    const keyring = privates
+      .get(this)
+      .keyring.getKeyringsByType(KeyringTypes.qr)[0];
+    return keyring || (await this.addQRKeyring());
```
Comment (gantunesr) [inline src/keyring/KeyringController.ts:636]: @soralit why is necessary to add a keyring instead of raising an error here?  I'm not sure if @mcmire already addressed this method and I'm missing something

File: src/keyring/KeyringController.ts (hunk 12 above)
Comment (gantunesr) [inline src/keyring/KeyringController.ts:454]: The method `signTypedMessage` is getting to big, in a next PR we should consider breaking it down to smaller internal methods

**Review by gantunesr — COMMENTED** [review]
I'm leaving some more comments to get the opinion of the other devs related to this development.

File: src/keyring/KeyringController.test.ts [hunk 14]
```diff
@@ -489,4 +500,344 @@ describe('KeyringController', () => {
...
+      const idBuffer = uuid.parse(requestId) as Uint8Array;
+      const ethSignature = new ETHSignature(
+        rlpSignatureData,
+        Buffer.from(idBuffer),
+      );
+      return ethSignature.toCBOR().toString('hex');
+    };
+
+    it('should sign message with QR keyring', async () => {
+      setTimeout(async () => {
```
[the remark this answers is not in the document]
Comment (soralit) [reply src/keyring/KeyringController.test.ts:578]: That's because QR keyring needs some user interaction to perform transaction signing and wallet syncing. So the implementation used an event emitter to wait for the QR code from Keystone side.

File: src/keyring/KeyringController.ts (hunk 10 above)
Comment (soralit) [reply src/keyring/KeyringController.ts:175]: Yes, I considered about the backward compatible before but sorry for forgot to mention it. This method is required to set different default names for QR keyring for UI requirement.

File: src/keyring/KeyringController.ts (hunk 13 above)
Comment (soralit) [reply src/keyring/KeyringController.ts:636]: I think @mcmire has a suggestion to this method before. 
The reason is that I want to make the method can always get the QR keyring instance so that we don't need to handle so many errors in the App layer.

File: src/keyring/KeyringController.ts (hunk 12 above)
Comment (soralit) [reply src/keyring/KeyringController.ts:454]: With the experience I have in MM extension integration before, I think the typed message signing can be hold by keyring implementations. Maybe it is better to create another controller called HardwareController later to handle things with hardwares that we cannot get the private key directly.

File: src/keyring/KeyringController.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply src/keyring/KeyringController.test.ts:578]: Hmm, I see. I'm wary of adding timeouts because I feel like they could lead to intermittently failing tests. This code will also run after the test is finished running, and if any part of it fails, it won't cause the test to fail (and Jest will give us a warning). Is there any way to get around this?

**Comment by gantunesr** [comment]
@soralit looks like your PR is not meeting the code coverage requirement

`Jest: "global" coverage threshold for functions (96%) not met: 95.94%
`

**Comment by mcmire** [comment]
@gantunesr The existing coverage is very close to the threshold, so any small changes could send it over the edge. I agree we should double-check the test coverage but I have a PR here that should make this better: https://github.com/MetaMask/controllers/pull/736.

File: src/keyring/KeyringController.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (soralit) [reply src/keyring/KeyringController.test.ts]: Hi @mcmire , thanks for your suggestion and I just stub some implementation with QR Keyring to write the unit tests so that we do not need setTimeout.

[the remark this answers is not in the document]
Comment (mcmire) [reply src/keyring/KeyringController.test.ts]: Thank you so much! That's much better :)

File: tests/setupTests.ts [hunk 15]
```diff
@@ -1 +1,9 @@
 import 'isomorphic-fetch';
+
+// some libraries (e.g. uuid) use defineProperty to export, so jest.mock and sinon.stub won't work.
```
[the remark this answers is not in the document]
Comment (soralit) [reply tests/setupTests.ts:3]: Yes. I tried it several times. With jest and sinon but none of them works.

[the remark this answers is not in the document]
Comment (soralit) [reply tests/setupTests.ts]: Found another way to mock the whole uuid package in `__mocks__/uuid.js`, and tested it worked well.

**Review by mcmire — DISMISSED** [review]
Thanks for putting up with all of our comments! I'm quite happy with this. Nice work!

**Review by gantunesr — DISMISSED** [review]
🚀 🚀 🚀 🚀

**Review by gantunesr — APPROVED** [review]
